src/ForSyDe/Shallow/MoC/CSDF.hs

```diff
@@ -0,0 +1,481 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  ForSyDe.Shallow.MoC.CSDF
+-- Copyright   :  (c) Ricardo Bonna, KTH/ICT/ES, ForSyDe-Group
+-- License     :  BSD-style (see the file LICENSE)
+--
+-- Maintainer  :  ricardobonna@gmail.com
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- Experimental lib. Further test needed
+--
+-----------------------------------------------------------------------------
+
+module ForSyDe.Shallow.MoC.CSDF (
+  -- -- * Combinational Process Constructors
+  -- -- | Combinational process constructors are used for processes that
+  -- -- do not have a state.
+  -- mapCSDF, zipWithCSDF, zipWith3CSDF, zipWith4CSDF,
+  -- * Sequential Process Constructors
+  -- | Sequential process constructors are used for processes that
+  -- have a state. One of the input parameters is the initial state.
+  delayCSDF, delaynCSDF,
+  -- -- * Processes
+  -- -- | Processes to unzip a signal of tupels into a tuple of signals
+  -- unzipCSDF, unzip3CSDF, unzip4CSDF,
+  -- * Actors
+  -- | Based on the process constructors in the CSDF-MoC, the
+  -- CSDF-library provides CSDF-actors with single or multiple inputs
+  actor11CSDF, actor12CSDF, actor13CSDF, actor14CSDF,
+  actor21CSDF, actor22CSDF, actor23CSDF, actor24CSDF,
+  actor31CSDF, actor32CSDF, actor33CSDF, actor34CSDF,
+  actor41CSDF, actor42CSDF, actor43CSDF, actor44CSDF
+  ) where
+
+import ForSyDe.Shallow.Core
+
+
+-------------------------------------
+--             --
+-- SEQUENTIAL PROCESS CONSTRUCTORS --
+--             --
+-------------------------------------
+
+-- | The process constructor 'delayCSDF' delays the signal one event
+--   cycle by introducing an initial value at the beginning of the
+--   output signal. Note, that this implies that there is one event
+--   (the first) at the output signal that has no corresponding event at
+--   the input signal. One could argue that input and output signals
+--   are not fully synchronized, even though all input events are
+--   synchronous with a corresponding output event. However, this is
+--   necessary to initialize feed-back loops.
+delayCSDF :: a -> Signal a -> Signal a
+delayCSDF x xs = x :- xs
+
+
+-- | The process constructor 'delaynCSDF' delays the signal n event
+--   cycles by introducing n initial values at the beginning of the
+--   output signal.
+delaynCSDF :: [a] -> Signal a -> Signal a
+delaynCSDF initial_tokens xs = signal initial_tokens +-+ xs
+
+
+------------------------------------------------------------------------
+--
+-- CSDF ACTORS
+--
+------------------------------------------------------------------------
+
+-- > Actors with one output
+
+-- | The process constructor 'actor11CSDF' constructs an CSDF actor with
```

Comment:
Document only one exported function per constructor. Provide in-line code examples in the documentation. The examples should be enough to see how to use "actors" in general, regardless of the number of inputs/outputs, so at least 2 examples should suffice. Point the other versions to the documented one (e.g. "See 'actor11'").